File: 3-2.php
<?php

$directions = file_get_contents('3.txt');

$length = strlen($directions);

$santa = ['x' => 0, 'y' => 0];
$robot = ['x' => 0, 'y' => 0];
$houses = ['0,0'];

for ($i = 0; $i < $length; ++$i) {
     ($i % 2 == 0 ? $position = &$santa : $position = &$robot);
    switch ($directions[$i]) {
        case '^':
            ++$position['y'];
            break;
        case 'v':
            --$position['y'];
            break;
        case '<':
            --$position['x'];
            break;
        case '>':
            ++$position['x'];
            break;
    }

    $house = implode(',', $position);

    if (!in_array($house, $houses)) {
        $houses[] = $house;
    }
}

var_dump(count($houses));

File: 3/src/bin/3-2.rs
use std::collections::HashSet;

fn main() {
    let directions = include_str!("3.txt").trim();

    let mut houses = HashSet::new();

    let mut santa = (0, 0);
    let mut robot = (0, 0);

    houses.insert(santa);

    for (index, direction) in directions.chars().enumerate() {
        let mut position = &mut santa;

        if index % 2 == 1 {
           position = &mut robot;
        }
        
        match direction {
            '^' => (*position).1 += 1,
            'v' => (*position).1 -= 1,
            '<' => (*position).0 -= 1,
            '>' => (*position).0 += 1,
            _ => println!("Weird direction!"),
        }
        houses.insert(*position);
    }

    println!("{}", houses.len());
}
